app/grandchallenge/reader_studies/forms.py

```diff
@@ -478,41 +480,174 @@ def clean_ground_truth(self):
         return values
 
 
-class DisplaySetForm(Form):
+class SelectUploadWidget(Select):
+    template_name = "reader_studies/select_upload_widget.html"
+
+
+class DisplaySetBaseForm(Form):
     _possible_widgets = {
-        Select,  # Default for ModelChoiceField
+        SelectUploadWidget,
         *InterfaceFormField._possible_widgets,
     }
+    user = None
 
-    def __init__(self, *args, instance=None, **kwargs):
+    def __init__(self, *args, instance, reader_study, user, **kwargs):
         super().__init__(*args, **kwargs)
-        if instance is not None:
-            for (
-                slug,
-                civ,
-            ) in instance.reader_study.values_for_interfaces.items():
-                val = instance.values.filter(interface__slug=slug).first()
-
-                if civ["kind"] in InterfaceKind.interface_type_json():
-                    # Use the field/widget provided by InterfaceFormField,
-                    # which includes proper validation
-                    self.fields[slug] = InterfaceFormField(
-                        kind=civ["kind"],
-                        schema=civ["schema"],
-                        initial=val.value if val else None,
-                        required=False,
-                    ).field
-                else:
-                    # Use a ModelChoiceField here, as InterfaceFormField would
-                    # provide an upload wodget, but we do not want to add new
-                    # images/files here, but rather assign existing values to
-                    # the proper display sets.
-                    self.fields[slug] = ModelChoiceField(
-                        queryset=ComponentInterfaceValue.objects.filter(
-                            id__in=civ["values"]
-                        ),
-                        initial=val,
-                        required=False,
-                    )
-
-            self.fields["order"] = IntegerField(initial=instance.order)
+        self.instance = instance
+        self.reader_study = reader_study
+        self.user = user
+        for slug, values in reader_study.values_for_interfaces.items():
+            current_value = None
+            if instance:
+                current_value = instance.values.filter(
+                    interface__slug=slug
+                ).first()
+            interface = ComponentInterface.objects.get(slug=slug)
```

Comment:
How many db queries is this doing? 2 per interface?

Reply:
Yes, 2 per interface for the update form and 1 per interface for the create form. They're small queries though, so I didn't think it would be a problem.